Integrations: Replace account IDs with display names in Jira comments.

reharsh opened the pull request:
This PR enhances the Jira integration by fetching account display names for mentions in comments using  
the Jira API. Account IDs are replaced with their corresponding display names to make Zulip notifications  
easier to read.  

Changes:  
- Update `view.py` to fetch account display names via the Jira API and replace `[~accountid: id]` with  
  the respective display names.  
- Document the process for generating `JIRA_API_TOKEN` in `/integrations/doc/jira`.  
- Add a new test case in `tests.py` to validate this functionality.  
- Add a new fixture `comment_created_with_account_mention.json` for testing account mentions.  

Fixes: #32646  

**Screenshots and screen captures:**
|                                    |                                     |
|-------------------------------------------|-------------------------------------------|
| Jira    | ![recent](https://github.com/user-attachments/assets/a8e6ceb7-70c7-486f-8fcf-84a1c64c7cb1)    |
| Zulip                          | ![Display Names](https://github.com/user-attachments/assets/7b2890e5-fbb2-463f-85de-01e682e0d06d)    |


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

reharsh commented:
Hi @PieterCK, this PR is up for review. Please take a look when you have a moment. Thanks!

PieterCK commented on zerver/webhooks/jira/doc.md:
```diff
@@ -14,6 +14,14 @@ Get Zulip notifications for your Jira projects!
+    If you don't do this, accounts mentioned in Jira comments will appear as
+    `[~accountid:6420c1740152b5f4f9f28a08]` in zulip, this is due to new api
+    changes in Jira.
+
```
In general, keep usually keep each line below 85 characters long, so make sure to enter a new line if a sentence gets too long.

You can make links into variables to help with this. For example, `slack/doc.md` does this.

reharsh replied:
done:)

PieterCK reviewed (commented):
Thanks for working on this! Did a quick documentation review, and left a couple of comments.

PieterCK commented:
Oh and the commit summary is a bit too long, I'd rephrase it to:
`integrations: Improve Jira's comment-event notification.`

reharsh commented:
Hi @PieterCK,  

Just a gentle reminder to review this PR when you have a chance. Your feedback would be greatly appreciated!  

Thank you!

PieterCK commented on zerver/webhooks/jira/view.py:
```diff
@@ -81,20 +85,63 @@ def convert_jira_markup(content: str, realm: Realm) -> str:
+
+def get_mentioned_account_name(url: str, auth: tuple[str, str], account_id: str) -> str:
+    r = requests.get(f"{url}?accountId={account_id}", auth=auth)
+    res = json.loads(r.text)["displayName"]
```
I think we want to do additional checks before assuming we got the response we want and handling those cases properly.

For reference:
```Python
def get_slack_api_data(slack_api_url: str, get_param: str, **kwargs: Any) -> Any:
    if not kwargs.get("token"):
        raise AssertionError("Slack token missing in kwargs")
    token = kwargs.pop("token")
    data = requests.get(slack_api_url, headers={"Authorization": f"Bearer {token}"}, params=kwargs)

    if data.status_code == requests.codes.ok:
        result = data.json()
        if not result["ok"]:
            raise Exception("Error accessing Slack API: {}".format(result["error"]))
        return result[get_param]

    raise Exception("HTTP error accessing the Slack API.")
```
One thing to note is, here, we need to take care to not throw Exception (error 500), instead we should probably return error 401 or 400 in the case where something unexpected happens.

PieterCK reviewed (commented):
@reharsh Thanks for making those changes! I did another review round on this, other than my comments below, I think we should try to use `config_options` for `jira_api_token` and `email`, take a look at #32797 for reference. I might do follow up review tomorrow.

Let me know if anything's unclear!

PieterCK commented on zerver/webhooks/jira/view.py:
```diff
@@ -81,20 +85,63 @@ def convert_jira_markup(content: str, realm: Realm) -> str:
+            else:
+                replacement = f"**{username}**"
+
+            content = content.replace(f"[~{username}]", replacement)
```
I gave this another look, we might be able to refactor `normalize_comment` into `convert_jira_markup` and just use that to reformat all event body that needs to be reformated, though I think they both are only used to reformat `commented` related events...

PieterCK commented on zerver/webhooks/jira/view.py:
```diff
@@ -78,23 +81,59 @@ def convert_jira_markup(content: str, realm: Realm) -> str:
+    # We replace mentioned account [~accountid:60e5a86a471e61006a4c51fd]
+    # with their corresponding usernames fetched by Jira API.
+    url = get_url(payload)
+    if auth:
```
Can this condition check for `url` too? if `get_url` returned `" "` then we can't make the API call too. I think that's worth putting for correctness, although in practice, the user's base URL looks like it's included in most payloads.

PieterCK commented on zerver/webhooks/jira/view.py:
```diff
@@ -78,23 +81,59 @@ def convert_jira_markup(content: str, realm: Realm) -> str:
+                replacement = f"**{username}**"
+            comment = comment.replace(f"{username}", replacement)
+        else:
+            raise SuspiciousOperation("Error accessing Jira API.")
```
I don't think we want to raise this error here. [SuspiciousOperation](https://docs.djangoproject.com/en/5.1/ref/exceptions/#suspiciousoperation) is raised when a user is performing a suspicious operation from security point of view.
>The [SuspiciousOperation](https://docs.djangoproject.com/en/5.1/ref/exceptions/#django.core.exceptions.SuspiciousOperation) exception is raised when a user has performed an operation that should be considered suspicious from a security perspective, such as tampering with a session cookie. Subclasses of SuspiciousOperation include:

We can use `JsonableError` from `zerver/lib/exceptions.py` instead.

I'm not sure why we need to do two replaces to reformat the username. 

So, all together I think we can adjust the function to this:
```python
# Process each match in the loop
    for match in matches:
        account_id = match.group(1)
        response = requests.get(f"{url}?accountId={account_id}", auth=auth)
        if response.status_code != requests.codes.ok:
            raise JsonableError(_("Error accessing Jira API."))
        data = response.json()
        username = data["displayName"]
        # We also try to convert a Jira user mention of format [~accountID:DUM1234] into a
        # Zulip user mention. We don't know the email, just the Jira username,
        # so we naively guess at their Zulip account using this
        user_profile = guess_zulip_user_from_jira(username, realm)
        if user_profile:
            username = user_profile.full_name
        comment = comment.replace(match.group(0), f"**{username}**")
    return comment
```

reharsh replied:
Thanks for the help @PieterCK !
I'll make the changes ASAP. also I just got to know that due to `config_options` email id is undergoing double URL encoding so my email : `reharshmeena@gmail.com` is transforming into `reharshmeena%2540gmail.com` which results into failed request, also trying to figure this out.

PieterCK replied:
Ah ok, we can also do that as a follow-up work if it's too messy to deal with now.

reharsh commented:
@PieterCK , I made the changes and also rephrased doc a bit for config_options, please take a look.

The reason for double URL encoding was because `URLSearchParams` in `integration_url_modal.ts` already encodes URL one time   then JIRA API call also encodes it one more time. so `@` becomes `%40` in integration URL then becomes `%2540` during request.

In this case I think better would be that we'll make integration_url's query params (config options) without any URL encoding e.g. `@` input as config option will be same as `@` in URL to be copied. I can put up separate PR for that, as it's not related to Jira integration. please let me know your thoughts, thanks!

PieterCK commented:
@reharsh I'm still not a 100% on using `config_options` with the issue on double encoded email in the URL param still persisting. I think it'd be great if there's a prep commit sorting that out first before we wholly commit to using `config_options` for the user's Jira email here. Opening a discussion in the #integrations channel first to get more input from other folks on your proposed fix you [commented](https://github.com/zulip/zulip/pull/32722#issuecomment-2565664504) on above might be worthwhile.

PieterCK commented:
On second thought, other integrations might not face the same situation, so I think we can leave `integration_url_modal.ts` to still encode stuff as is and just add an extra decoding step in the Jira integration for the `email` param.  The `%2540` string once decoded the first time around will become just `%40`, which if we decode again will revert back to `@`.

PieterCK commented on zerver/webhooks/jira/doc.md:
```diff
@@ -14,6 +14,12 @@ Get Zulip notifications for your Jira projects!
 
 1. {!generate-webhook-url-basic.md!}
 
+      To include a mentioned user's display name in your Jira notifications,
```
Just a small nit on phrasing!

```
1. {!generate-webhook-url-basic.md!}
   By default, mentioned users in your Jira comment will appear as
   `[~accountid:MENTIONED_ACC_ID]` in Zulip. To get a mentioned
   user's display name in your Zulip notifications, you can fill the
   **Your Jira API token** and **Your Jira email** fields when
   creating the **integration URL**. For instructions on generating
   your `JIRA_API_TOKEN`, refer to this [Jira documentation][1].
```

PieterCK commented on zerver/webhooks/jira/view.py:
```diff
@@ -78,21 +86,52 @@ def convert_jira_markup(content: str, realm: Realm) -> str:
+    # Use re.finditer to find all matches
+    matches = list(pattern.finditer(comment))
+
+    # Process each match in the loop
```
I don't think we need this code comment and the one above (line 188) as that is obvious.

reharsh commented:
@PieterCK, I've made the changes! can you please take a look again:)

PieterCK reviewed (changes requested):
@reharsh I've left some more comments, should have made them on the previous review round, sorry! Once you've got the chance to address them, this basically LGTM. 
 
I've also done some end-to-end testing with the latest group of suggested changes:
![image](https://github.com/user-attachments/assets/7948f4c1-fff5-4dd4-a1b6-fc2f04070be2)

Oh, I think the PR descriptions should also contain some screenshots of the visual changes introduced by the PR, such as the integration doc and the generat integration URL modal for Jira.

PieterCK commented on zerver/webhooks/jira/doc.md:
```diff
@@ -14,6 +14,13 @@ Get Zulip notifications for your Jira projects!
 
 1. {!generate-webhook-url-basic.md!}
 
```
we want to delete this newline as per the suggestion I commented [here](https://github.com/zulip/zulip/pull/32722#discussion_r1903111607), if not the whole paragraph won't get indented correctly 

![image](https://github.com/user-attachments/assets/01b23058-12b4-4b3d-a6df-d28f8d035241)

reharsh commented:
> @reharsh I've left some more comments, should have made them on the previous review round, sorry! Once you've got the chance to address them, this basically LGTM.

No worries! I'll make the changes ASAP, also thanks for commit suggestions:)

reharsh commented:
hii @PieterCK , done with the changes! I've used `urlsplit` instead of `urlparse` because there was linting guideline suggesting this, also I forgot to mention earlier, we are also decoding Jira_api_token because there were similar issues with that too when I tested, in case you were wondering. PTAL

PieterCK commented:
Thanks for making those changes, this LGTM. 

Notes for next reviewers:
- I guess this is preexisting, but I wonder if we really need `guess_zulip_user_from_jira`. If the user's full name is their Zulip username, then there's no need to query for their profile.
- A full check up of this integration might be worth considering. It seems like we’re maintaining backward compatibility in a way that feels a bit hacky. The most Frankenstein-ish is probably `handle_updated_issue_event`, it has some logic there that covers the commented events even though we have dedicated functions for those sets of events. 

Let me know if we want to turn any of these into follow-up issue. Though I guess this PR might be small enough to sneak in the smaller fix.

@reharsh Feel free to call out any issues you notice while working on this

PieterCK commented:
@laurynmm -- I'm tagging this as ready for maintainer review! I've done some review and tested this end-to-end. Posted some review notes [here](https://github.com/zulip/zulip/pull/32722#issuecomment-2582479190). 

@zulipbot add "maintainer review"

reharsh commented:
Hii @laurynmm!
Just curious to know if you've had a chance to review this PR. Please let me know if there’s anything you'd like me to update.

laurynmm commented on zerver/webhooks/jira/view.py:
```diff
@@ -369,8 +418,16 @@ def api_jira_webhook(
+    auth = None
+
+    if email and jira_api_token is not None:
+        # Config options encode parameters using Jira's internal encoding,
```
This comment makes no sense as config options are part of Zulip's Webhook integration code ...

laurynmm commented on zerver/webhooks/jira/view.py:
```diff
@@ -369,8 +418,16 @@ def api_jira_webhook(
+        # Config options encode parameters using Jira's internal encoding,
+        # causing double encoding. To address this, we decode the parameters
+        # to remove the extra layer of encoding.
+        auth = (unquote(email), unquote(jira_api_token))
```
Let's use a dataclass here instead of a tuple.

laurynmm commented on zerver/webhooks/jira/view.py:
```diff
@@ -44,10 +52,12 @@ def guess_zulip_user_from_jira(jira_username: str, realm: Realm) -> UserProfile
 
 
-def convert_jira_markup(content: str, realm: Realm) -> str:
+def convert_jira_markup(payload: WildValue, realm: Realm, auth: tuple[str, str] | None) -> str:
```
I would pull the changes to use this function for the comment events in to a separate commit. Probably before adding the config parameters.

reharsh replied:
done!

laurynmm commented on zerver/webhooks/jira/view.py:
```diff
@@ -78,21 +88,52 @@ def convert_jira_markup(content: str, realm: Realm) -> str:
+) -> Any:
+    response = requests.get(jira_api_url, auth=auth, params=kwargs)
+    if response.status_code != requests.codes.ok:
+        raise JsonableError(_("Error accessing jira api."))
```
Do we want the notification message to fail in this case? Maybe we could just offer a nicer version of the Jira account ID in that case.

Also, Jira and API should be capitalized as such in the current error message.

reharsh replied:
how is this?
| jira | zulip |
| --- | --- |
| ![Before Image](https://github.com/user-attachments/assets/10041316-d878-4a98-a741-745eafdf0cea) | ![After Image](https://github.com/user-attachments/assets/2a4a60b1-310a-44a1-ad3b-252090224015) |

alya replied:
Maybe just "Unknown Jira user" or something? I don't think that ID is super helpful.

laurynmm reviewed (commented):
@reharsh - Thanks for working on this issue! I did an initial pass on these changes and had some overall feedback about how to improve these changes a git. Let me know if you have any questions about my comments.

I haven't yet looked at the documentation changes, but can do that when this initial feedback has been addresssed. Thanks again!

reharsh committed:
integration: Refactor use of convert_jira_markup in comment events.

reharsh committed:
integrations: Improve Jira's "comment"-event notification.

This commit improves the Jira integration by fetching Display Names
for accounts mentioned in comments using the Jira API. Account IDs in
comments are replaced with their corresponding Display Names to make
Zulip notifications more readable.

Fixes: #32646

reharsh commented:
hello @laurynmm, made the changes, please check.

reharsh commented:
Hi, @laurynmm, @alya. Now, the message shows "unknown Jira user (858342098098...)." I think it'll be very rare for multiple users to have the same number, and this is still better than the default or simple counter.
<img width="713" alt="Screenshot 2025-01-23 at 6 47 00 PM" src="https://github.com/user-attachments/assets/730f95ad-2938-4455-87e3-ae8f633bc141" />

PS. I'll be mostly out for few days so please let me know if any more changes are required to this so I can try to make this pr mergeable since it is on release board, thanks!

alya commented:
That looks OK to me. I guess the next step is to make the PR  pass tests. Thanks!

reharsh commented:
pushed again, not sure why tests failed earlier.
